grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java

```diff
@@ -75,104 +74,94 @@
 
             this.delegate = delegate;
             try {
-                for (Consumer<Listener<I>> task : pendingTasks) {
-                    task.accept(delegate);
+                for (;;) {
+                    inLoop = true;
```

Comment:
Initially, I thought of this style but this could break the order of execution.
Let's say:

- pending_queue: [method1, method2, method3]
- method2 invokes method4
- I would expect the methods to be called in order but the actual order is `method1 -> method2 -> method4 -> method3`. I don't see this is an expected situation.

Reply:
So I intended to allocate a new array and then a new task to it rather than executing it directly.

Reply:
> executing it directly. 

which part do you mean by?
before entering into the loop and after exiting the loop, events can be run immediately. IMO
and it's event-driven and the state might be handled well so I think events can be executed async without consideration of the order.
Hmm

Reply:
I think I'm missing something. let me draw a state diagram.

Reply:
I meant this.
https://github.com/line/armeria/pull/4729/files#diff-42092cd0259a3e7d170dd151d5f9d6924ba51c556ac691ef4408461c9275069eR144

> it's event-driven and the state might be handled well so I think events can be executed async without consideration of the order.

I don't think event-driven implies the order is not considered. `onReady` should be called before invoking `onMessage()`. `onHalfClose()` or `onComplete()` will follow `onMessage()`.

> executing it directly.

If `onMessage()` invokes `onComplete()` while `onHalfClose()` is in the pending queue, originally, I thought `onComplete()` needs to be called after  `onHalfClose()`. That said, it is fine to invoke ``onComplete()` before `onHalfClose()` because the original non-async version of `Interceptor` has the same order.

Reply:
That means the current implementation looks good to me.

Reply:
> I don't think event-driven implies the order is not considered. onReady should be called before invoking onMessage(). onHalfClose() or onComplete() will follow onMessage().

yes. IMHO, order of `onReady`,  `onMessage()`. `onHalfClose()` or `onComplete()` is fixed by state machine of grpc.
but By `responseObserver`,  `onNext`, `onCompleted`, and `onError` can be called from the user's codes anytime. I think if a request is canceled by timeout and `responseObserver` is passed to the async process. then those events still can be called if the timing is not good. and the method can call below even after already been canceled or closed.
`io.grpc.stub.ServerCall`
<img width="1076" alt="image" src="https://user-images.githubusercontent.com/322936/227420015-57979e3e-90d9-4f67-9c07-b0ad3a772585.png">
but. `onNext` handle well by looking at the current state(state machine)